drivers/sensor/bq34110/bq34110.c

```diff
@@ -0,0 +1,371 @@
+		return -ENODEV;
+	}
+
+	ret = bq34110_ctrl_reg_read(dev, SUB_DEVICE_TYPE,
+				    &device_type, MAC_DATA);
+	if (ret < 0) {
+		LOG_ERR("Failed to get device type");
+		return -EIO;
+	}
+
+	if (device_type != BQ34110_DEVICE_ID) {
+		LOG_ERR("Invalid Device");
+		return -EINVAL;
+	}
+
+	memset(&writedata, 0, sizeof(writedata));
+	design_voltage = bq34110_cfg->design_voltage;
+	writedata[0] = (design_voltage >> 8) & 0x00FF;
+	writedata[1] = (design_voltage & 0xFF);
+
+	/* Update Design Voltage */
+	ret = updateDFparameter(dev, DESIGN_VOLTAGE, writedata, sizeof(writedata));
+	if (ret < 0) {
+		LOG_ERR("Failed to update design voltage");
+		return ret;
+	}
+
+	k_msleep(BQ34110_DELAY);
+
+	memset(&writedata, 0, sizeof(writedata));
+	design_capacity = bq34110_cfg->design_capacity;
+	writedata[0] = (design_capacity >> 8) & 0x00FF;
+	writedata[1] = (design_capacity & 0xFF);
+
+	/* Update Design Capacity */
+	ret = updateDFparameter(dev, DESIGN_CAPACITY_MAH, writedata,
+				sizeof(writedata));
+	if (ret < 0) {
+		LOG_ERR("Failed to update Design Capacity");
+		return ret;
+	}
+
+	k_msleep(BQ34110_DELAY);
+
+	memset(&writedata, 0, sizeof(writedata));
+	writedata[0] = 0x02;
+	writedata[1] = 0x04;
+
+	/* Select Internal Temperature Sensor */
+	ret = updateDFparameter(dev, OPERATION_CONFIG_A, writedata,
+				sizeof(writedata));
+	if (ret < 0) {
+		LOG_ERR("Failed to update Operation Config A");
+		return ret;
+	}
+
+	memset(&writedata, 0, sizeof(writedata));
+	taper_current = bq34110_cfg->taper_current;
+	writedata[0] = (taper_current >> 8) & 0x00FF;
+	writedata[1] = (taper_current & 0xFF);
+
+	/* Update Taper Current */
+	ret = updateDFparameter(dev, TAPER_CURRENT, writedata, sizeof(writedata));
+	if (ret < 0) {
+		LOG_ERR("Failed to update Taper Current");
+		return ret;
+	}
+
+	k_msleep(BQ34110_DELAY);
+
+	memset(&writedata, 0, sizeof(writedata));
+	taper_voltage = bq34110_cfg->taper_voltage;
+	writedata[0] = (taper_voltage >> 8) & 0x00FF;
+	writedata[1] = (taper_voltage & 0xFF);
+
+	/* Update Taper Voltage */
+	ret = updateDFparameter(dev, TAPER_VOLTAGE, writedata, sizeof(writedata));
+	if (ret < 0) {
+		LOG_ERR("Failed to update Taper Voltage");
+		return ret;
+	}
+
+	k_msleep(BQ34110_DELAY);
+
+	return 0;
+}
+
+static const struct sensor_driver_api bq34110_battery_driver_api = {
+	.sample_fetch = bq34110_sample_fetch,
+	.channel_get = bq34110_channel_get,
+};
+
+static struct bq34110_data bq34110_data;
+
+static struct bq34110_config bq34110_cfg = {
+	.i2c_addr = DT_INST_REG_ADDR(0),
+	.i2c_dev = DEVICE_DT_GET(DT_INST_BUS(0)),
+	.taper_current = DT_INST_PROP(0, taper_current),
+	.taper_voltage = DT_INST_PROP(0, taper_voltage),
+	.design_voltage = DT_INST_PROP(0, design_voltage),
+	.design_capacity = DT_INST_PROP(0, design_capacity),
+};
+
+DEVICE_DT_INST_DEFINE(0, &bq34110_gauge_init, device_pm_control_nop,
+		      &bq34110_data, &bq34110_cfg, POST_KERNEL,
+		      CONFIG_SENSOR_INIT_PRIORITY,
+		      &bq34110_battery_driver_api);
```

Comment:
Please extend this to support multiple instances by using `DT_INST_FOREACH_STATUS_OKAY`